### PR TITLE
Simple websocket client

### LINKED ISSUE #12: Create basic websocket client
It's time to write code!
We need to understand the [Boost.Beast examples](https://www.boost.org/doc/libs/1_72_0/libs/beast/doc/html/beast/examples.html#beast.examples.clients) and write a simple websocket client. I suggest writing an asynchronous version right away. The host and the message will be passed through command line arguments.

The client main goals:
- Send requests;
- Receive requests.

@c71n93 @temikfart any additions?

### PR DESCRIPTION
**Pull request by cann3v**
Coroutines based websocket client.
Executable takes 3 arguments: host, port, message. For example `./coolstory-gram-client localhost 8080 "Hello, world!"`.
This arguments passed into `void run(...)`, within which an event loop is started (`ioc.run()`).
Function `do_session(...)` performs some actions:
- Resolves remote host;
- Connects on the IP address;
- Performs handshake;
- Sends the message;
- Receives the message.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -10,7 +10,7 @@ int main(int argc, char** argv)
                   "Usage: coolstory_gram_client <host> <port> <text>\n" <<
                   "Example:\n" <<
                   "    coolstory_gram_client echo.websocket.org 80 \"Hello, world!\"\n";
-        return EXIT_FAILURE;
+        return EXIT_SUCCESS;
```
Comment (temikfart) [inline src/main.cpp]: Need to fix build scripts, but not the code.
Change run command and add `--help` flag. Of course add `--help` flag processing in the app.

File: CMakeLists.txt [hunk 2]
```diff
@@ -21,5 +21,6 @@ FetchContent_MakeAvailable(json logger)
 include_directories(include)
 
 add_executable(coolstory_gram_client
-            src/main.cpp)
-target_link_libraries(coolstory_gram_client PRIVATE nlohmann_json::nlohmann_json liblogger)
+            src/main.cpp
+			src/websocket_session.cpp)
```
Comment (temikfart) [inline CMakeLists.txt:26]: Tabulation?

File: CMakeLists.txt [hunk 3]
```diff
@@ -21,5 +21,6 @@ FetchContent_MakeAvailable(json logger)
...
+target_link_libraries(coolstory_gram_client PRIVATE nlohmann_json::nlohmann_json liblogger boost_coroutine)
```
Comment (temikfart) [inline CMakeLists.txt]: Boost::coroutine like in CSGServer.
Don't forget to fix build script:
- Add `include_directories` for Boost headers;
- Add `REQUIRED COMPONENTS coroutine`.
See `CMakeLists.txt` in the CSGServer.

File: src/main.cpp [hunk 4]
```diff
@@ -1,7 +1,22 @@
...
 
-int main() {
-    std::cout << "Hello, world!" << std::endl;
+int main(int argc, char** argv)
+{
+    logger::init(logger::debug, logger::cout);
 
-    return 0;
+    if(argc != 4)
+    {
```
Comment (temikfart) [inline src/main.cpp]: Spaces!
Open curly brace must be in the same line as operator.

File: src/main.cpp [hunk 5]
```diff
@@ -1,7 +1,22 @@
...
+        std::cerr <<
+                  "Usage: coolstory_gram_client <host> <port> <text>\n" <<
+                  "Example:\n" <<
+                  "    coolstory_gram_client echo.websocket.org 80 \"Hello, world!\"\n";
+        return EXIT_SUCCESS;
```
Comment (temikfart) [inline src/main.cpp:12]: `EXIT_FAILURE` more preferable.

File: src/websocket_session.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if(ec)
+        return fail(ec, "resolve");
+
+    // Set a timeout on the operation
+    beast::get_lowest_layer(ws).expires_after(std::chrono::seconds(30));
+
+    // Make the connection on the IP address we get from a lookup
+    auto ep = beast::get_lowest_layer(ws).async_connect(results, yield[ec]);
+    if(ec)
+        return fail(ec, "connect");
```
Comment (temikfart) [inline src/websocket_session.cpp]: `if (ec)` - fix spacing.

File: src/websocket_session.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Read a message into our buffer
+    ws.async_read(buffer, yield[ec]);
+    if(ec)
+        return fail(ec, "read");
+
+    // Close the WebSocket connection
+    ws.async_close(websocket::close_code::normal, yield[ec]);
+    if(ec)
+        return fail(ec, "close");
```
Comment (temikfart) [inline src/websocket_session.cpp]: Spaces!

**Review by temikfart — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 1 above)
Comment (cann3v) [reply src/main.cpp]: We will count the number of arguments instead of parsing them.

**Review by temikfart — APPROVED** [review]

**Pull request merged**